src/main/kotlin/com/example/ipwhitelist/repository/AppRepository.kt

```diff
@@ -0,0 +1,261 @@
+        val queryConditional = sortBeginsWith(
+            Key.builder()
+                .partitionValue(AppTableKeyPrefix.APP.prefix + appKey)
+                .sortValue(AppTableKeyPrefix.USER.prefix)
+                .build()
+        )
+
+        val queryEnhancedRequest = QueryEnhancedRequest.builder()
+            .queryConditional(queryConditional)
+            .filterExpression(
+                Expression.builder()
+                    .expression("#role = :role")
+                    .putExpressionValue(":role", AttributeValue.builder().s("ADMIN").build())
+                    .expressionNames(Collections.singletonMap("#role", "role"))
+                    .build()
+            )
+            .build()
+
+        val appAdminsIterator = appUsersTable.query(queryEnhancedRequest)
+
+        val appAdmins: MutableList<String> = ArrayList()
+        appAdminsIterator.items().forEach { appAdmins.add(it.objectId.substringAfter(AppTableKeyPrefix.USER.prefix)) }
+
+        return appAdmins
+    }
+
+    fun deleteByAppId(appKey: String) : Boolean {
+        val appDetailsTable = getTable(ApplicationDetails::class.java)
+        val appUsersTable = getTable(ApplicationUser::class.java)
+
+        val appDetailsQueryRequest = QueryEnhancedRequest.builder()
+            .queryConditional(
+                sortBeginsWith(
+                    Key.builder()
+                        .partitionValue(appKey)
+                        .sortValue(AppTableKeyPrefix.APP.prefix)
+                        .build()
+                )
+            )
+            .build()
+
+        val appDetailsToDelete = appDetailsTable.query(appDetailsQueryRequest).items()
+
+        val appDetailsDeleteRequests = appDetailsToDelete.map {
+            TransactDeleteItemEnhancedRequest.builder()
+                .key(
+                    Key.builder()
+                        .partitionValue(it.appId)
+                        .sortValue(it.objectId)
+                        .build()
+                )
+                .build()
+        }
+
+        val appUsersQueryRequest = QueryEnhancedRequest.builder()
+            .queryConditional(
+                sortBeginsWith(
+                    Key.builder()
+                        .partitionValue(appKey)
+                        .sortValue(AppTableKeyPrefix.USER.prefix)
+                        .build()
+                )
+            )
+            .build()
+
+        val appUsersToDelete = appDetailsTable.query(appUsersQueryRequest).items()
+        val appUsersDeleteRequests = appUsersToDelete.map {
+            TransactDeleteItemEnhancedRequest.builder()
+                .key(
+                    Key.builder()
+                        .partitionValue(it.appId)
+                        .sortValue(it.objectId)
+                        .build()
+                )
+                .build()
+        }
+
+        try {
+            dynamoDbEnhancedClient.transactWriteItems {
```

Comment:
I think calling transactWriteItems twice will result in 2 seperate transactions, you can add both appDetailsRequests and appUsersRequests into one transactWriteItems call

Reply:
You are right, fixed!